### PR TITLE
Avoid race in accessing nvidiaDevices between Setup() and GetCollector()

### PR DESCRIPTION
**Pull request by mindprince**
Fixes https://github.com/google/cadvisor/issues/1810

### CODE REVIEW

**Comment by mindprince** [comment]
This PR is just fixing the case where we could have a concurrent map access.

The time taken by `initializeNVML()` on  my desktop is in ms but on some cloud environments it can take up to 5s. That's why I didn't put it in the critical path. We can see how big of an issue this is and change that in the future. My guess is that it's unlikely that there are containers with GPUs attached when the kubelet starts up but I maybe wrong and we can change that in the future.

**Comment by choury** [comment]
> it's unlikely that there are containers with GPUs attached when the kubelet starts up

We should not require to restart all containers if we just want to restart or upgrade kubelet.
May be call  GetCollector after NVML initialized or wait it in GetCollector?

**Comment by mindprince** [comment]
> We should not require to restart all containers if we just want to restart or upgrade kubelet.

Agreed. Good point.

> May be call GetCollector after NVML initialized or wait it in GetCollector?

Not sure what you mean exactly. Can you elaborate?

**Comment by dashpole** [comment]
> My guess is that it's unlikely that there are containers with GPUs attached when the kubelet starts up but I maybe wrong and we can change that in the future.

will this work across kubelet restarts then?

**Comment by vishh** [comment]
LGTM

**Comment by mindprince** [comment]
Discussed with @vishh and made changes to work better in case of kubelet restarts.

**Comment by vishh** [comment]
LGTM again. @dashpole WDYT?

File: accelerators/nvidia.go [hunk 1]
```diff
@@ -129,7 +139,9 @@ func (nm *NvidiaManager) Destroy() {
 // present in the devices.list file in the given devicesCgroupPath.
 func (nm *NvidiaManager) GetCollector(devicesCgroupPath string) (AcceleratorCollector, error) {
 	nc := &NvidiaCollector{}
+	nm.RLock()
 	if !nm.nvmlInitialized || len(nm.nvidiaDevices) == 0 {
+		nm.RUnlock()
```
Comment (dashpole) [inline accelerators/nvidia.go]: do we want to defer unlock here?  What if we do not enter into this if statement?

Comment (mindprince) [reply accelerators/nvidia.go]: Updated. Thanks for noticing this! (facepalm)

**Pull request merged**